src/v/archival/ntp_archiver_service.cc

```diff
@@ -1121,30 +1316,36 @@ ss::future<ntp_archiver::batch_result> ntp_archiver::wait_all_scheduled_uploads(
           return s.upload_kind == segment_upload_kind::non_compacted;
       });
 
+    // Inline manifest upload in regular non-compacted uploads
+    // if any were scheduled.
+    bool inline_manifest_in_non_compacted_uploads = false;
+    for (const auto& i : non_compacted_uploads) {
+        if (i.result) {
+            inline_manifest_in_non_compacted_uploads = true;
+            break;
+        }
+    }
```

Comment:
Yeah, this is a pre-existing quirk.  The code that awaits the segments also walks the list to extract just the ones with  `result` set.  I didn't fully unpick the earlier logic that identifies segments to uploads but doesn't necessarily kick them all off, it can probably be simplified.